tensorboard/plugins/hparams/tf_hparams_query_pane/tf-hparams-query-pane.html

```diff
@@ -580,29 +635,66 @@
         this.set("_metrics", result);
       },
 
-      // Computes the _schema property from the _experiment and _metrics
-      // properties.
       _computeSchema() {
-        // We have to use Polymer's 'path'-based array mutations so that
-        // other elements (like a dom-repeat template) that depend on
-        // schema.metricInfos and schema.hparamInfos would be notified of the
-        // change and update correctly.
+        return {
+          hparamColumns: this._hparams.map(
+              hparam => ({hparamInfo: hparam.info,
+                          displayed: hparam.displayed})),
+          metricColumns: this._metrics.map(
+              metric => ({metricInfo: metric.info,
+                          displayed: metric.displayed})),
+          visibleSchema: this._computeVisibleSchema(),
+          loaded: true,
+        };
+      },
+
+      _hparamVisibilityChanged(event) {
+        const hparamIndex = event.model.index;
+        // We'd like the notifications for the changed schema to be sent to
+        // consuming elements after both this.schema.hparamColumns and
+        // schema.visibleSchema are changed so that the schema is in a coherent
+        // state. Thus we first silently modify hparamColumns,
+        // then visibleSchema--which will also send the notification for the
+        // visibileSchema change--and then call notifyPath to send the
+        // notification for the hparamColumns change.
+        this.schema.hparamColumns[hparamIndex].displayed =
+            this._hparams[hparamIndex].displayed;
+        this._updateVisibleSchema();
+        this.notifyPath(`schema.hparamColumns.${hparamIndex}.displayed`);
+      },
+
+      _metricVisibilityChanged(event) {
+        const metricIndex = event.model.index;
+        // We'd like the notifications for the changed schema to be sent to
+        // consuming elements after both this.schema.metricColumns and
+        // schema.visibleSchema are changed so that the schema is in a coherent
+        // state. Thus we first silently modify metricColumns,
+        // then visibleSchema--which will also send the notification for the
+        // visibileSchema change--and then call notifyPath to send the
+        // notification for the metricColumns change.
+        this.schema.metricColumns[metricIndex].displayed =
+            this._metrics[metricIndex].displayed;
+        this._updateVisibleSchema();
+        this.notifyPath(`schema.metricColumns.${metricIndex}.displayed`);
```

Comment:
I believe if you use `set` API, you don't need to notify like this.
```ts
        this.set(`schema.metricColumns.${metricIndex}.displayed`, this._metrics[metricIndex].displayed);
```

(Also, instead of imperatively calling `updateVisibleSchema`, make sure there is an observer that, on metricColumn display changes, trigger `updateVisibleSchema`.

Similar concern above for `_hparamVisibilityChanged`.

Reply:
The issue here is with atomically modifying schema so that observing elements will see the changes to both visibleSchema and metricColumns together. If I use the 'set' API then the notification for the changed metricColumns property will go out immediately, and an observing element will see a schema object in an incosistent state--one whose visibleSchema does not match the new 'metricColumn' information. This may create problems: for example, it would make the assumptions that the code in tf.hparam.utils.getAbsoluteColumnIndex() not hold.
I know this is not ideal, but I don't know of a better solution. I tried to explain that in the comment above the code. I'm open to suggestions.

Reply:
> While your rationale makes sense, it looks rather brittle or unconventional. How about this?

I agree. I don't like it very much either, and I'd be happy with a better solution.
Using: 'this.schema = this._computeSchema()' will replace the identity of the schema object each time the user changes the visibility of a single column. This makes it harder for for observing elements to redraw only the changed part. See also the comments in lines 345--350. For example, if the schema object stays the same and we only change its properties, the vaadin-grid element can set the column property  'hidden' to be linked to the (negation of) the corresponding 'displayed' subproperty of schema.hparamColumn or schema.metricColumn.

Reply:
If the point of the PR is a refactor, I would advise against such performance optimization (without carefully looking at what is costing the most CPU cycle; i.e., premature optimization). Also, there are many other ways to mitigate such issue while keeping your code sensible. 

One small gotcha with this approach was, which I found while poking around with your branch, with current code, both [1] and [2] while [3] does not (missing `*`; i.e., `schema.*`. Without [3] getting triggered, InteractionManager _can_ go out of sync with the `options` which fortuitously did not happen because we were mutating schema by reference. However, while I could not trigger it, I think they could really go out of sync soon (depending on execution order) when [4] happens. The point here is that, while the bug I pointed out _is_ solvable, it just makes everything more brittle and harder to reason without concerning the greater system. 

[1]: https://github.com/erzel/tensorboard/blob/par_coords_refactor/tensorboard/plugins/hparams/tf_hparams_scale_and_color_controls/tf-hparams-scale-and-color-controls.html#L184
[2]: https://github.com/erzel/tensorboard/blob/par_coords_refactor/tensorboard/plugins/hparams/tf_hparams_parallel_coords_plot/tf-hparams-parallel-coords-plot.html#L227-L230
[3]: https://github.com/erzel/tensorboard/blob/c8808d369b1fbb79f270e0053b86becbced82632/tensorboard/plugins/hparams/tf_hparams_parallel_coords_plot/tf-hparams-parallel-coords-plot.html#L197
[4]: https://github.com/erzel/tensorboard/blob/c8808d369b1fbb79f270e0053b86becbced82632/tensorboard/plugins/hparams/tf_hparams_query_pane/tf-hparams-query-pane.html#L562